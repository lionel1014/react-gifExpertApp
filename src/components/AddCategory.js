import React, { useState } from 'react'
import PropTypes from 'prop-types'


export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');  // => ('') > () => ('') != () => ( ) da una error, envia una variale undefined

    const handleInputChange = (e) =>{//los valores que recibimos del input
        const {value} = e.target;
        setInputValue ( value );

        // console.log('handleInputChange fue llamado con exito')

    };

    const handleSubmit = (e) =>{//para enviar el input en un form
        e.preventDefault();
        
        if ( inputValue.trim().length >= 3 ){
            setCategories(cate => [inputValue,...cate]); // guardamos el nuevo valor del `inputValue`
            setInputValue (''); // el estado del inputValue ahora es vacío
        }

    }

    return (
        <form onSubmit = {handleSubmit} > 
            <p> {inputValue} </p>
            <input type="text" value={inputValue} placeholder='Ingresa una categoría'
            onChange = { handleInputChange }
            />
        </form>
    )
}

AddCategory.propTypes = {

    setCategories : PropTypes.func.isRequired

};
