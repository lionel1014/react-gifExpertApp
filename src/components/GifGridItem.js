import React from 'react';
import PropTypes from 'prop-types'; 

export const GifGridItem = ({title,url}) => {

    // console.log({title,url})

    return (
        <div className="imagenes animate__animated animate__flipInX" >
            <p> {title} </p>
            <img src={ url } alt= {title} />
        </div>
    )
}

GifGridItem.propTypes ={

    title : PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};
