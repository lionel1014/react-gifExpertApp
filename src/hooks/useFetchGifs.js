import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';


//un custom hook
export const useFetchGifs = (category) => {
    
    const [objeto, setObjeto] = useState({
        data : [],
        loading : true,    
    });

    useEffect( () => {
        getGifs(category)
            .then(imgs => {

                // console.log(imgs);
                setObjeto({ 
                    data : imgs,
                    loading : false
                });

            })
    }, [category])

    // a los 4  seg setea Objeto con este "objeto" (valga la redundancia)
    // setTimeout(() => {
    //     setObjeto({
    //         data : [1,2,3,4,5,6,10],
    //         loading : false,    
    //     });
    // }, 4000);

    return objeto; // {data = [], loading = true}

};
