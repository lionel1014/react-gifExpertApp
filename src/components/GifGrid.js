import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => { // no te olvides de desestructurar la propiedad

    const {data:images,loading} = useFetchGifs(category);  //data:images, en la desestructuracion cambiamos el nombre de la variable asi
    
    // useEffect(() =>{
    //     //importamos la promesa getGifs de "helpers"
    //     getGifs(category)
    //         .then(setImages)
    //         // .then( imgs => setImages(imgs)) // una funcion que manda el primer argumento al primer argumento de setImages
    //         // se lo puede resumir como el de arriba
    // },[category]); // que haga lo que hay dentro si cambia category

    return (
        <>
        <h3 className="animate__animated animate__bounceInLeft "> {category} </h3>
        {/* {loading ? 'Cargando...' : 'Fin de la carga del objeto.'} */}
         {loading && <p className="animate__animated animate__bounceIn " >Cargando...</p>}  {/*el && es una condición lógica AND */}
        
        <div className="imagenesGrid" >
            
            {
                images.map( ( img ) =>{
                    return <GifGridItem 
                            key = {img.id}
                            {...img} 
                            />
                }) 
            }
            
        </div>
        </>
    )
}
