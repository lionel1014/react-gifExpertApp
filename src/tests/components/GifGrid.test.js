import React from 'react'
import { shallow } from 'enzyme'
import { GifGrid } from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');


describe('Pruebas en el componente <GifGrid/>', () => {

    const category = 'Pepe';
    // let wrapper = shallow( <GifGrid category={category} /> )

    // beforeEach( () =>{    
    //     wrapper = shallow( <GifGrid category={category} /> )
    // });

    //SNAPSHOT
    test('Debe mostrarse correctamente (Snapshot)', () => {
        
        //cuando se llame esta funcion, regresar estos valores
        useFetchGifs.mockReturnValue({
            data : [],
            loading : true
        }); 

        const wrapper = shallow( <GifGrid category={category} /> )
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe mostrar ítems cuando se cargan imágenes con useFetchGifs(category)', () => {
        
        const gifs =[{
            id : '123',
            url : 'https//localhost/cosa/mavale.jpg',
            title : 'Pepito',
        },
        {
            id : 'abc',
            url : 'https//localhost/cosa/mavale.jpg',
            title : 'Pepapig',
        }];

        useFetchGifs.mockReturnValue({
            data : gifs,
            loading : false
        }); 

        const wrapper = shallow( <GifGrid category={category} /> );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false); // pregunto si es parrafo existe, la carga tiene que ser falsa, o sea trajo algun elemento
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length) // pregunto por la cantidad de elementos, si es igual a la cantidad de objetos
    });
        
});
