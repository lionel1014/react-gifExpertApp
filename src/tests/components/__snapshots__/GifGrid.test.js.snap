// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrid/> Debe mostrar ítems cuando se cargan imágenes con useFetchGifs(category) 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounceInLeft "
  >
     
    Pepe
     
  </h3>
    
  <div
    className="imagenesGrid"
  >
    <GifGridItem
      id="123"
      key="123"
      title="Pepito"
      url="https//localhost/cosa/mavale.jpg"
    />
    <GifGridItem
      id="abc"
      key="abc"
      title="Pepapig"
      url="https//localhost/cosa/mavale.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en el componente <GifGrid/> Debe mostrarse correctamente (Snapshot) 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounceInLeft "
  >
     
    Pepe
     
  </h3>
  <p
    className="animate__animated animate__bounceIn "
  >
    Cargando...
  </p>
    
  <div
    className="imagenesGrid"
  />
</Fragment>
`;
